feat(docs): add first explain blog post (#1325)
questdb/questdb.io: blog/2023-01-26-introducing-explain.mdx

SQL is just the query interface and the database must decide on a great many
things, before it can produce results, such as:

- How to access a table - e.g. full or index scan, scan forward or backward.
- How to join tables - e.g. nested loop, merge or hash join.
- How to order joins - e.g. 5 tables can be joined in 5! = 120 different orders.
- How to optimize expressions - e.g. evaluate constant expression once instead
  of for every row.

The above is just the tip of the iceberg, a lot is happening under the hood and
many factors influence the end performance of a query.

## `EXPLAIN` explained

[EXPLAIN](https://questdb.io/docs/reference/sql/explain) SQL keyword shows the
execution plan (the outcome of step 3. above), which is the logical algorithm
used to implement a given statement and thus 'explains' its performance. Even
though it's not part of the SQL standard, many databases, e.g. PostgreSQL, MySQL or
Oracle, do implement their own variant of `EXPLAIN`. Contrary to SQL, which is a
high-level declarative language, `EXPLAIN` uses, often low-level, terms related
to database implementation, for example, hash joins or index scan. Learning

> these [concepts](https://questdb.io/docs/concept/storage-model/#read-model) is

Reviewer: I am linking it to docs/introduction/concepts/!

essential to fully understanding how queries work and how to improve them.

It was thanks to this type of tool that I was able to first, understand what and
why happens under the hood, and second - reduce nightly processes run time well
beyond what was expected.

Now, it's time for practice. Let's have a look at the following queries that use
the trades table [in the QuestDB demo instance](https://demo.questdb.io/):

```questdb-sql
CREATE TABLE trades (
  symbol SYMBOL CAPACITY 256 CACHE,
  side SYMBOL CAPACITY 256 CACHE,
  price DOUBLE,
  amount DOUBLE,
  timestamp TIMESTAMP
) TIMESTAMP (timestamp) PARTITION BY DAY;
```

Query A:

```questdb-sql
SELECT count(*) FROM trades WHERE price = 2.1f;
```

Query B:

```questdb-sql
SELECT count(*) FROM trades WHERE price = 2.1;
```

> For hot data response times are consistent:

Reviewer: I think this is just for my own understanding: why these are consistent? Query A tends to take longer than B in a consistent way?

Author: Removed the word.  I meant to say that times a 'stable' .

- Query A: 160ms
- Query B: 50ms

Can you guess why?

<Banner
  alt="Taking a quiz."
  height={433}
  src="/img/blog/2023-01-26/quiz-time.jpg"
  width={650}
>
  Photo by <a href="https://unsplash.com/@nguyendhn">Nguyen Dang Hoang Nhu</a>{" "}
  via on
  <a href="https://unsplash.com">Unsplash</a>
</Banner>

It could be a number of things, but why guess when you can check:

```questdb-sql

> EXPLAIN SELECT count(*) FROM trades WHERE price = 0.0f;

Reviewer: Query A is not 0.0f but 2.1f, does this matter?
```suggestion
EXPLAIN SELECT count(*) FROM trades WHERE price = 2.1f;
```

Author: Fixed.

| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: price=0.0f                                                    |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame                                                 |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan (2)              |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades (1) |

The output is a tree of nodes - stages in query processing - that have
properties and/or subnodes (aka child nodes).

In the plan above we have the following nodes and properties:

- nodes - `Count`, `Async Filter`, `DataFrame`, `Row forward scan`,
  `Frame forward scan` . They start with upper-case letter and have bigger
  indents than properties .
- properties - `price=0.0f`and `on: trades .`

According to the plan, the query contains the following steps:

1. Forward scan all the partitons of the table.
1. Forward scan all rows in each partition.
1. Apply filters to each row.
1. Count all rows that pass it.

Next, we check the plan for the second query :

```questdb-sql

> EXPLAIN SELECT count(*) FROM trades WHERE price = 0.0;

Reviewer: Again if we refer to query B why is it 0.0 here?
```suggestion
EXPLAIN SELECT count(*) FROM trades WHERE price = 2.1;
```

Author: Fixed.

```

| QUERY PLAN                                                                                            |
| ----------------------------------------------------------------------------------------------------- |
| Count                                                                                                 |
| &nbsp;&nbsp;&nbsp;&nbsp;Async JIT Filter                                                              |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: price=0.0d                                                |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame                                             |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan              |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: trades |

Can you spot the difference?

It's about the **_JIT_**.

The second query is faster because it uses

> [Just-In-Time-compiled](/blog/2022/01/12/jit-sql-compiler/) filter!

Reviewer: ```suggestion
[Just-In-Time-compiled](/docs/concept/jit-compiler/) filter!
```
I think using Doc is better than a blog post?

Author: Quality-wise blog post is better but doc does refer to it so I think it's fine and applied your suggestion .

The first query doesn't use it due to column and argument type mismatch.

In both cases, the `trades` table is fully scanned, which is reflected by
`DataFrame` and the child nodes.

:::warning

The QuestDB Web Console limits the number of output rows to 1k, so it is not the
best way to measure the performance of queries that return data sets bigger than
the limit.

:::

Let's look at the `pos`, another table available
[in the QuestDB demo instance](https://demo.questdb.io/):

```questdb-sql
CREATE TABLE pos (
  time TIMESTAMP,
  id SYMBOL CAPACITY 256 CACHE INDEX CAPACITY 512,
  lat DOUBLE,
  lon DOUBLE,
  geo6 GEOHASH(6c),
  geo12 GEOHASH(12c)
) TIMESTAMP (time) PARTITION BY DAY;
```

What if we need to count the number of records with a given `id` prefix ?

```questdb-sql
SELECT count(*)
FROM pos
WHERE substring(id,1,21) = 'NSTTQFPKISYEGDLPJFEKP'
678
```

Response time - 1.2s.

A bit slow for just a few hundred records.

> Let's inspect the plan:

Reviewer: Should we add the explain query here?

Author: It'd simply repeat the query above with explain prefix, so a bit verbose and not really offer much considering that there's a number of exaplain examples earlier  .  What do you think ?

| QUERY PLAN                                                                                         |
| -------------------------------------------------------------------------------------------------- |
| Count (1)                                                                                          |
| &nbsp;&nbsp;&nbsp;&nbsp;Async Filter (2)                                                           |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: substring(id,1,21)='NSTTQFPKISYEGDLPJFEKPS'            |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;workers: 14                                                    |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame (3)                                      |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan           |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos |

Query counts (1) records filtered by asynchronous, multi-threaded filter (2),

> which scans the whole table (3).

Reviewer: the ref for (3) is unclear to me because you did not mention DataFrame...

Author: I added (3) to both subnodes of DataFrame .

> Query counts (1) records filtered by asynchronous, multi-threaded filter (2),
> which scans the whole table (3).

Reviewer: ```suggestion
The query counts (1) records filtered by an asynchronous, multi-threaded filter (2),
which scans the whole table (3).
```

So now I understand what the numbers are referring to. The current sentence does not convert a sequence. It describes a nesting relationship, but I wonder if the number corresponds to the sequence of the operation? Does it mean that it follows the following steps:
1. record count
2. apply filter
3. scan table

I don't think so and I think the steps are in reverse. Should we use a number list to describe the steps and number the them differently? As:
1. A full table scan is triggered.
2. An asynchronous, multi-threaded filter is applied to the record.
3. Filtered records are counted.
And we number the plan accordingly too...

Maybe I get it al wrong...

Author: These numbers were meant to point to pieces of execution plan, not show the sequence of operation .
You're right about the actual steps.

Considering the number of records (678 vs 120M in the table) it seems strange
that the index on id wasn't used.

What can we do here?

In many databases, index access requires a query predicate to match the
expression used in the index definition.

If there's a mismatch, then the index would need to be fully scanned, which is
much more costly than a single value or value range lookup.

In this case, we have a `symbol` column, so assuming that `id` contains A-Z
letters only and is 22-characters long, we can rewrite the expression with a
`IN[to add link]` clause :

```questdb-sql
SELECT count(*)
FROM pos
WHERE id IN ('NSTTQFPKISYEGDLPJFEKPA','NSTTQFPKISYEGDLPJFEKPB' ... 'NSTTQFPKISYEGDLPJFEKPZ');
```

> In effect, the response time shrinks to 3ms, about 400 times.

Reviewer: 400 times less?

Author: Fixed.

where timestamp &gt; '2023-01-16T09:51:43.000000Z'
```

This time result arrives in 1 millisecond. What's going on? `EXPLAIN` shows :

| QUERY PLAN                                                                                                            |
| --------------------------------------------------------------------------------------------------------------------- |
| Count                                                                                                                 |
| &nbsp;&nbsp;&nbsp;&nbsp;DataFrame                                                                                     |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan                                                      |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interval forward scan on: trades                                      |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;intervals: [static=[1673862703000001,9223372036854775807] |

Instead of a full scan we've only an interval scan limited to
['2023-01-16T09:51:43.000000Z', max timestamp] time range.

Now, we've got the right plan but don't know the reason why the database chooses

> it over the former one. To understand the difference we've to take a look at the
> documentation of the
> [now() function](https://questdb.io/docs/reference/function/date-time/#now):

Reviewer: ```suggestion
it over the former one. To understand the difference, we need to understand the `sysdatetime()` function: `sysdatetime()` changes within the query execution timeframe.
```

I am guess from the context that the first query uses sysdatetime() which is costly. The function now() is not really mentioned above so using it could be confusing...

Author: I agree . Maybe it'd be better to add the warning to systimestamp() and sysdate() functions and then use excerpt from the former here ?

> Unlike sysdatetime(), it does not change within the query execution timeframe
> and should be used in WHERE clause to filter designated timestamp column
> relative to current time

> 

Reviewer: Could we not copy from the Docs as things might change. Maybe rephrase this?

learn some fundamental concepts. That's why, before going any further, let's
focus on basic stuff for a moment.

QuestDB processes queries in the following phases :

1. Parsing - Query text is parsed and a model is created.
2. Optimization - The query is rewritten for the best performance.
3. Planning - An execution plan is created, and the best table scan and join
   methods are chosen.
4. Execution - Tables are accessed according to the execution plan and outcomes
   are produced.

![Diagram showing the phases of processing a query](/img/blog/2023-01-26/query-phases.png)

SQL is just the query interface and the database must decide on a great many
things before it can produce results. These things include:

- How to access a table - e.g. full or index scan, scan forward or backward.
- How to join tables - e.g. nested loop, merge, or hash join.
- How to order joins - e.g. 5 tables can be joined in 5! = 120 different orders.
- How to optimize expressions - e.g. evaluate constant expression once instead
  of for every row.

The above is just the tip of the iceberg, a lot is happening under the hood and
many factors influence the end performance of a query.

## `EXPLAIN` explained

> The [EXPLAIN](https://questdb.io/docs/reference/sql/explain) SQL keyword shows

Reviewer 2: ```suggestion
The [EXPLAIN](https://questdb.io/docs/reference/sql/explain/) SQL keyword shows
```

| ------------------------------------------------------------------------------------------------------ |
| Count (1)                                                                                              |
| &nbsp;&nbsp;&nbsp;&nbsp;Async Filter (2)                                                               |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;filter: substring(id,1,21)='NSTTQFPKISYEGDLPJFEKPS'                |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;workers: 14                                                        |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DataFrame (3)                                          |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan (3)           |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos (3) |

I have annotated the plan with some numbers to illustrate the information. From
the table, we can see that query counts (1 in the table) records filtered by an
asynchronous, multi-threaded filter (marked as 2), which scans the whole table
(marked as 3).

> Considering the number of records (678 rows in the result v.s. 120M in the whole

Reviewer: ```suggestion
Considering the number of records (678 rows in the result vs 120M in the whole
```

| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Row forward scan (3)           |
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Frame forward scan on: pos (3) |

Query counts (1) records filtered by asynchronous, multi-threaded filter (2),
which scans the whole table (3).
I have annotated the plan with some numbers to illustrate the information. From
the table, we can see that query counts (1 in the table) records filtered by an
asynchronous, multi-threaded filter (marked as 2), which scans the whole table
(marked as 3).

Considering the number of records (678 vs 120M in the table) it seems strange

> that the index on `id` wasn't used.

Author: v.s. -> vs. ?
